src/Servers/Kestrel/test/Interop.FunctionalTests/HttpClientHttp2InteropTests.cs

```diff
@@ -0,0 +1,561 @@
+                    {
+                        var reader = context.Request.BodyReader;
+
+                        var readResult = await reader.ReadAsync();
+                        while (!readResult.IsCompleted && readResult.Buffer.Length < "Hello World".Length)
+                        {
+                            reader.AdvanceTo(readResult.Buffer.Start, readResult.Buffer.End);
+                            readResult = await reader.ReadAsync();
+                        }
+
+                        var sequence = readResult.Buffer.Slice(0, "Hello World".Length);
+                        Assert.True(sequence.IsSingleSegment);
+                        await context.Response.BodyWriter.WriteAsync(sequence.First);
+                        reader.AdvanceTo(sequence.End);
+                        await context.Features.Get<IHttpResponseCompletionFeature>().CompleteAsync();
+
+                        try
+                        {
+                            // The client sends one more packet after the server completes
+                            readResult = await reader.ReadAsync();
+                            while (!readResult.IsCompleted && readResult.Buffer.Length < "Hello World".Length)
+                            {
+                                reader.AdvanceTo(readResult.Buffer.Start, readResult.Buffer.End);
+                                readResult = await reader.ReadAsync();
+                            }
+
+                            Assert.True(readResult.Buffer.IsSingleSegment);
+                            var result = Encoding.UTF8.GetString(readResult.Buffer.FirstSpan);
+                            reader.AdvanceTo(readResult.Buffer.End);
+
+                            var finalResult = await reader.ReadAsync();
+                            Assert.True(finalResult.IsCompleted && finalResult.Buffer.Length == 0);
+                            lastPacket.SetResult(result);
+                        }
+                        catch (Exception ex)
+                        {
+                            lastPacket.SetException(ex);
+                        }
+                    }));
+                });
+            using var host = await hostBuilder.StartAsync();
+
+            var url = host.MakeUrl(scheme);
+
+            using var client = CreateClient();
+            client.DefaultRequestHeaders.ExpectContinue = true;
+
+            var streamingContent = new StreamingContent();
+            var request = CreateRequestMessage(HttpMethod.Post, url, streamingContent);
+            var responseTask = client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead).DefaultTimeout();
+            // The server doesn't respond until we send the first set of data
+            await streamingContent.SendAsync("Hello World").DefaultTimeout();
+            var response = await responseTask;
+
+            Assert.Equal(HttpVersion.Version20, response.Version);
+            var stream = await response.Content.ReadAsStreamAsync();
```

Comment:
We should probably be putting DefaultTimeout() on all awaits in the main test method to be safe. I know it's annoying, but I think that should include stuff like HostBuilder.StartAsync() too.